### ISSUE TITLE
Feature Request: Add discard option to all crypttab entries generated by coreos installer

### ISSUE THREAD

**Issue by N3rdle** [coreos/fedora-coreos-tracker#1392]
Hi!

I request that those most familiar with the plumbing used to generate the crypttab file (and/or cryptsetup commands) used during the initramfs as well as during the later stages of boot of a persisted coreos operating system allow end-users to include some way to add the "discard" option to each entry created for an SSD in the crypttab file during installation.

As currently implemented, end-users can change the crypttab file on an installed operating system instance (for example in the post-install script, among other ways), but such out-of-band solutions feel fragile and undesirable, and especially so because the root filesystem itself gets mounted by the initramfs, and as a result our changes in the root filesystem's crypttab cannot ever affect the root filesystem itself.

For those of us that have SSDs behind lux in coreos, none of the blocks therefore ever able to be informed for discard (TRIM) to the underlying storage devices which is likely to lead to substantially worse performance (and possibly some longevity impacts) over time.

**Comment by bgilbert** [comment]
Yeah, I agree.  I think most users will probably want this, for hardware longevity, but it does have security consequences.  Discard makes it possible to tell which parts of the disk are currently used and thus make some guesses about what's stored on it.  If we were implementing LUKS support from scratch, we might have opted to enable discard by default nevertheless, but it's harder to justify making that change behind the backs of existing configs.

So in sum I think this would need a new field in the Ignition spec, defaulting to our current behavior.

**Comment by N3rdle** [comment]
Roger. Your approach makes total sense to me. I myself think that security
concern is an acceptable risk for many (most) use cases, and one I’ll take
most of the time given my threat models.

If you’re going to go through that effort you might want to consider
simultaneously adding a mechanism to turn off the read and write queues for
luks volumes on ssds as well.

Everything I’m seeing says it makes a considerable improvement in io
throughput and latency with the queues off.

Thanks for your consideration here!

Cheers

Peter

**Comment by bgilbert** [comment]
Implementation inhttps://github.com/coreos/ignition/pull/1540.

**Comment by dustymabe** [comment]
Ignition v2.15.0 was released!

**Comment by dustymabe** [comment]
The fix for this went into `next` stream release `38.20230310.1.0`. Please try out the new release and report issues.

**Comment by dustymabe** [comment]
The fix for this went into `testing` stream release `37.20230303.2.0`. Please try out the new release and report issues.

**Comment by dustymabe** [comment]
The fix for this went into `stable` stream release `37.20230303.3.0`.
